src/Matcher.js

```diff
@@ -0,0 +1,116 @@
+import isEqual from 'lodash/isEqual';
+import pathToRegexp from 'path-to-regexp';
+
+export default class Matcher {
+  constructor(routeConfig) {
+    this.routeConfig = routeConfig;
+  }
+
+  match({ pathname }) {
+    const matches = this.matchRoutes(this.routeConfig, pathname);
+    if (!matches) {
+      return null;
+    }
+
+    const routeIndices = new Array(matches.length);
+    const routeParams = new Array(matches.length);
+    const params = {};
+
+    matches.forEach((routeMatch, i) => {
+      routeIndices[i] = routeMatch.index;
+      routeParams[i] = routeMatch.params;
+      Object.assign(params, routeMatch.params);
+    });
+
+    return { routeIndices, routeParams, params };
+  }
+
+  matchRoutes(routeConfig, pathname) {
+    for (let index = 0; index < routeConfig.length; ++index) {
+      const route = routeConfig[index];
+
+      const match = this.matchRoute(route, pathname);
+      if (!match) {
+        continue; // eslint-disable-line no-continue
+      }
+
+      const { children = [] } = route;
+      const { params, remaining } = match;
+
+      const childMatches = this.matchRoutes(children, remaining);
+      if (childMatches) {
+        return [{ index, params }, ...childMatches];
+      }
+
+      if (!remaining) {
```

Comment:
i might just be dense here, but when would a route match, not match any children, and have  a path `remaining`?